src/Datadog.Trace/DiagnosticListeners/AspNetCoreDiagnosticObserver.cs

```diff
@@ -240,6 +229,41 @@ private void OnHostingUnhandledException(object arg)
                 _options.OnError?.Invoke(span, exception, httpContext);
             }
         }
+
+        private readonly struct HeadersCollectionAdapter : IHeadersCollection
+        {
+            private readonly IHeaderDictionary _headers;
+
+            public HeadersCollectionAdapter(IHeaderDictionary headers)
+            {
+                _headers = headers;
+            }
+
+            public IEnumerable<string> GetValues(string name)
+            {
+                if (_headers.TryGetValue(name, out var values))
+                {
+                    return values.ToArray();
```

Comment:
This `ToArray()` allocates another string array and copy the content (https://source.dot.net/#System.Linq/EnumerableHelpers.Linq.cs,105) maybe we can avoid it by using:

```suggestion
                    return values as string[] ?? values.ToArray();
```

what do you think?

Reply:
This is `StringValues.ToArray()`, not `Enumerable.ToArray()` (confusing, I know)
https://source.dot.net/#Microsoft.Extensions.Primitives/StringValues.cs,f7f329a587dc1a5e

Reply:
ok, yeah didn't realize that, this struct is only used by the `HttpRequest.Headers`